runtime-sdk/modules/rofl-market/src/policy.rs

```diff
@@ -0,0 +1,123 @@
+use std::collections::BTreeMap;
+
+use base64::prelude::*;
+
+use oasis_runtime_sdk::{
+    core::{
+        common::crypto::signature::{PublicKey, Signature},
+        consensus::registry::{EndorsedCapabilityTEE, Node},
+        host::attestation::{LabelAttestation, ATTEST_LABELS_SIGNATURE_CONTEXT},
+    },
+    modules::rofl::{
+        policy::{AllowedEndorsement, EndorsementPolicyEvaluator},
+        Error,
+    },
+    types::address::Address,
+    Context,
+};
+
+use crate::{state, types::InstanceId};
+
+/// Name of the ROFL app instance metadata key used to store the provider attestation.
+pub const METADATA_KEY_POLICY_PROVIDER_ATTESTATION: &str = "net.oasis.policy.provider";
+/// Name of the provider label set by the scheduler.
+pub const LABEL_PROVIDER: &str = "net.oasis.provider";
+
+/// Value of the `LABEL_PROVIDER` label as set by the scheduler.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct ProviderLabel {
+    /// Address of the provider.
+    pub provider: Address,
+    /// Instance identifier.
+    pub instance: InstanceId,
+}
+
+/// Provider attestation metadata stored in `METADATA_KEY_POLICY_PROVIDER_ATTESTATION` label.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct ProviderAttestation {
+    /// A CBOR-serialized `LabelAttestation`.
+    pub label_attestation: Vec<u8>,
+    /// Signature from endorsing node.
+    pub signature: Signature,
+}
+
+/// An endorsement policy evaluator that supports provider-related constraints via lookups
+/// in ROFL market module state.
+pub struct ProviderEndorsementPolicyEvaluator;
+
+impl EndorsementPolicyEvaluator for ProviderEndorsementPolicyEvaluator {
+    fn verify_atom<C: Context>(
+        _ctx: &C,
+        policy: &AllowedEndorsement,
+        ect: &EndorsedCapabilityTEE,
+        endorsing_node_id: PublicKey,
+        _endorsing_node: &Option<Node>,
+        metadata: &BTreeMap<String, String>,
+    ) -> Result<(), Error> {
+        match policy {
+            AllowedEndorsement::Provider(address) => {
+                // Check if the endorsing node is endorsed by the given provider.
+                let provider = state::get_provider(*address).ok_or(Error::NodeNotAllowed)?;
+                if !provider.nodes.contains(&endorsing_node_id) {
+                    return Err(Error::NodeNotAllowed);
+                }
+                Ok(())
+            }
+            AllowedEndorsement::ProviderInstanceAdmin(expected_admin) => {
+                let pa: ProviderAttestation = cbor::from_slice(
+                    &BASE64_STANDARD
+                        .decode(
+                            metadata
+                                .get(METADATA_KEY_POLICY_PROVIDER_ATTESTATION)
+                                .ok_or(Error::NodeNotAllowed)?,
```

Comment:
Will this be hard to debug since all errors are the same?

Reply:
That's a good point, but the problem is that this can be part of a more complex condition and it would only be a single basic atom failing. So even if this fails, the overall policy may still evaluate to success.